Partly refactor and nerf some traps (caltrops, nailboards, glass)

RenechCDDA opened the pull request:
#### Summary
Balance "Nerf some traps(caltrops, nailboards, glass). Larger creatures can destroy them with their weight, or otherwise render them ineffective and need to be setup again"

#### Purpose of change
I've read some stories of people cheesing hulks or other mindless enemies (without trapavoid) by luring them back and forth over a single caltrop or nailboard. 

As you all know: No fun allowed.

#### Describe the solution
Rewrite most of the old dusty code to make rational assumptions. For example a quadruped player that steps on a caltrop will no longer stop, pause, and then roll around to make sure all their legs and arms properly fall victim to the same caltrop. Only one limb used for movement will be affected, picked at random. (Also removed the hardcoded checks for left/right leg)

Since a tile is approximately 1x1 meters across, scaled the chance to actually run into the caltrops/board/glass trap by the creature's size, instead of being guaranteed when moving into that tile. The caltrop trap is a single caltrop, the nailboard trap is three boards and 20 nails or so. There's just not enough surface area to get anything close to 100% coverage unless the creature is freakishly huge.

Made metal caltrops and nailboards sometimes get destroyed by the triggering creature, scaling chance based on the creature's weight. A zombie behemoth the size of a truck should just roll over the damn thing

Even if the trap isn't destroyed, any triggering can sometimes "disarm" those two traps, rendering them back into item form. This would be the equivalent of e.g. the nailboard getting turned over. Caltrops are typically designed so that they always rest with a point upwards, but they could be bent out of shape or partly embedded into the ground at an angle that makes them ineffective.

Equalized the damage for these traps when triggering vs monsters as opposed to triggering vs characters. No reason for it to do more or less damage depending on the target.

Fix a call to make sound in the glass caltrops function which only fired if the player was able to see it when it triggered
![image](https://github.com/user-attachments/assets/8e53a8d7-0a22-4a70-aabb-d2f0ad163e7c)

Simplified and replaced some old no-longer-necessary dynamic casts to monster pointers with is_monster() and the like

Safety-checked a cast to character by making sure the cast doesn't result in nullptr before calling it

Changed a numeric value from being stored as `char`(????????????) to int

https://github.com/CleverRaven/Cataclysm-DDA/blob/master/src/trapfunc.cpp#L155

#### Describe alternatives you've considered
JSONizing traps??

#### Testing
Spawned some lines of caltrops, glass, and nailboards. Lead a bunch of fat zombies back and forth over them. Lines of traps weren't guaranteed to kill zombies at any depth, and the horde left the field of traps pretty sparse after a few passes. Still a lot of zombies died and there was blood everywhere, which is about what I'd expect for a deep line of traps.

#### Additional context

RenechCDDA committed:
Partly refactor and nerf some traps (caltrops, board, glass)

Night-Pryanik commented on src/trapfunc.cpp:
```diff
@@ -222,56 +219,74 @@ bool trapfunc::beartrap( const tripoint &p, Creature *c, item * )
+        // destroy trap
+        here.remove_trap( p );
+        if( !c->is_avatar() ) {
+            // stupid branch to properly capitalize the start of the sentence
```
Why don't you use `disp_name` for both cases? `disp_name( false, true )` for monsters will give you `The <monster> destroys...`. `disp_name( false, true )` for NPCs will give you `John Doe destroys...`

RenechCDDA replied:
Simple answer: I found out afterwards and hadn't bothered to change it

![image](https://github.com/user-attachments/assets/025cea3a-3695-4ca9-bbf7-6788d5b0abd1)

Night-Pryanik commented on src/trapfunc.cpp:
```diff
@@ -312,28 +344,22 @@ bool trapfunc::caltrops_glass( const tripoint &p, Creature *c, item * )
-            c->deal_damage( nullptr, bp, damage_instance( damage_cut, rng( 9, 30 ) ) );
-        }
+    if( c->is_monster() ) {
+        add_msg_if_player_sees( p, _( "The %s steps on a sharp glass caltrop!" ), c->get_name() );
```
No other traps affected by your PR, including metal caltrop, gets this line. Is it really needed? If yes, maybe add it to other traps as well, for consistency?

RenechCDDA replied:
Apparently add_msg_player_or_npc and <npcname> handles monsters as an overload, but without proper capitalization

![image](https://github.com/user-attachments/assets/2e07a9b7-f206-4fd8-9b73-9bd221acdfda)

More refactor, more refactor..

Night-Pryanik commented:
While you're here, could you please add a case of player character's mount stepping on a glass caltrop? Metal caltrops process this case, while glass caltrops don't.

RenechCDDA commented:
All good points, let me get to them soon(tm)

RenechCDDA commented:
![image](https://github.com/user-attachments/assets/3e542991-7023-49df-baa5-46113d6a3b8a)

The horse stuff actually needed a bit (more...) refactoring, because (friendly) NPCs can ride horses too.

RenechCDDA committed:
More trapfunc refactor
-Uniformly apply the mount moves loss
-Characters stepping onto these traps without a mount lose moves too, not just zeds
-Fix the display messages when monsters step on them
-Separate out the mount messages and tetanus application as static functions

ADekema commented:
The caltrops probably need a bit of a rework. They most likely wouldn´t stay in place that long on account of getting stuck in some creatures feet. Making the trap dissapear but also imposing a steep movement and some accuracy and dodge penalty on any creature for as long as it´s stuck. Maybe we could reuse the code that makes arrows be stuck into an enemy to make caltrops stuck into a creature.
